lib/rubygems/core_ext/kernel_warn.rb

```diff
@@ -36,20 +28,23 @@ class << self
             break
           end
 
-          start += 1
-
           if path = loc.path
             unless path.start_with?(rubygems_path) or path.start_with?('<internal:')
               # Non-rubygems frames
               uplevel -= 1
+              break if uplevel < 0
             end
           end
+
+          start += 1
         end
-        uplevel = start
+        kw[:uplevel] = start
       end
 
-      kw[:uplevel] = uplevel
-      original_warn.call(*messages, **kw)
-    }
+      super(*messages, **kw)
+    end
   end
+
+  extend Gem::KernelExt
+  Kernel.singleton_class.prepend Gem::KernelExt
```

Comment:
I think this will not work for a non-main-object non-`Kernel.warn` receiver, like if you call `warn` in some method of some class.

Reply:
Could you add a test for that to your PR, to prove that your approach works better? Thanks!

Reply:
Confirmed it doesn't work in this case, in that that would call the original Kernel#warn and skip RubyGems' variant.
Trying to write a test for it, it's not easy.